### PR TITLE
add GroupedNormalNormal distribution

### PR DESCRIPTION
**Pull request by martinjankowiak**
Conjugate distribution for Normally distributed data that is controlled by a latent mean that is also controlled by a Normal distribution. Supports ragged grouped data.

### CODE REVIEW

File: pyro/distributions/grouped_normal_normal.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        obs_scale_sq_inv_sum = torch.zeros_like(self.prior_loc).scatter_add(
+            0, self.group_idx, obs_scale_sq_inv
+        )
+        precision = prior_scale_sq_inv + obs_scale_sq_inv_sum
+        scaled_value_sum = torch.zeros_like(self.prior_loc).scatter_add(
+            0, self.group_idx, value * obs_scale_sq_inv
+        )
+
+        loc = (scaled_value_sum + self.prior_loc * prior_scale_sq_inv) / precision
+        scale = precision.pow(-0.5)
```
Comment (fritzo) [inline pyro/distributions/grouped_normal_normal.py:108]: nit: I believe `.rsqrt()` is cheaper

File: pyro/distributions/grouped_normal_normal.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        self.num_data_per_batch = prior_loc.new_zeros(self.num_groups).scatter_add(
+            0, self.group_idx, prior_loc.new_ones(self.group_idx.shape)
+        )
+        super().__init__(batch_shape, validate_args=validate_args)
+
+    def expand(self, batch_shape, _instance=None):
+        raise NotImplementedError
+
+    def sample(self, sample_shape=()):
+        raise NotImplementedError
```
Comment (fritzo) [inline pyro/distributions/grouped_normal_normal.py:88]: Gosh it seems a shame not to implement at least a generator for test data
```python
has_rsample = True
def rsample(self, sample_shape=torch.Size()):
    group_part = Normal(self.prior_loc, self.prior_scale).rsample(sample_shape)
    obs_part = Normal(0, self.obs_scale).rsample(sample_shape)
    return group_part[..., self.group_idx] + obs_part
```
but I guess it's fine to omit it until someone needs it 🤔

File: tests/distributions/test_grouped_normal_normal.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+import math
+
+import torch
+
+from pyro.distributions import GroupedNormalNormal, Normal
+from tests.common import assert_close
+
+
+def test_grouped_normal_normal(num_groups=3, num_samples=10 * 6):
```
Comment (fritzo) [inline tests/distributions/test_grouped_normal_normal.py:12]: did you mean `10 ** 6`?  would that allow tighter `atol` below?

File: pyro/distributions/grouped_normal_normal.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        "prior_scale": constraints.positive,
+        "obs_scale": constraints.positive,
+    }
+    support = constraints.real
+
+    def __init__(
+        self, prior_loc, prior_scale, obs_scale, group_idx, validate_args=None
+    ):
+        prior_loc, prior_scale = broadcast_all(prior_loc, prior_scale)
+        obs_scale, group_idx = broadcast_all(obs_scale, group_idx)
```
Comment (fritzo) [inline pyro/distributions/grouped_normal_normal.py:62]: Could you add some shape assertions to clarify requiremets?  I'm guessing:
```python
assert prior_loc.dim() == 1
assert obs_scale.dim() == 1
```

**Review by fritzo — COMMENTED** [review]
Looks good, just a couple minor comments

File: pyro/distributions/grouped_normal_normal.py (hunk 2 above)
Comment (martinjankowiak) [reply pyro/distributions/grouped_normal_normal.py:88]: yeah the thing is the only distribution that a user is actually likely to care about is the posterior predictive and not the prior predictive

File: tests/distributions/test_grouped_normal_normal.py (hunk 3 above)
Comment (martinjankowiak) [reply tests/distributions/test_grouped_normal_normal.py]: good catch 👍

**Comment by martinjankowiak** [comment]
thanks should be good to go now 👍

**Review by fritzo — APPROVED** [review]

**Pull request merged**